Add throttling to postinstall

MrQubo opened the pull request:
Installing any global packages is currently VERY slow.

It works by spawning the process in the background on every execution of the hook. It executes the reshim only if it's the last hook call within the last 10 seconds.

I've tested this one on Arch Linux. It works well and reduces the execution times for individual hook to the minimum.

You can test it by adding `echo ASDF RESHIM NODEJS` above `asdf reshim` command. Then execute `npm i -g npm`, it will call the hook a few hundred times, but the echo message should be displayed only once (with 10-second delay after last hook call).

MrQubo commented:
I think the delay should be reduced. 2 seconds or even 1 should work perfectly.

MrQubo commented:
Actually `npm i -g npm` will spawn hooks for dependencies only if the npm itself will be updated. You can try with some other package with a lot of dependencies like `npx`. Just uninstall it if you have it installed and install it again.

MrQubo commented on bin/postinstall:
```diff
@@ -1,3 +1,42 @@
 #!/usr/bin/env bash
 
-asdf reshim nodejs "$ASDF_INSTALL_VERSION"
+set -e
```
We might consider dropping this line.

Stratus3D replied:
I like being strict when it comes to scripts like this.

Stratus3D commented on bin/postinstall:
```diff
@@ -1,3 +1,48 @@
+set -e
+
+declare LOCK_FILE="/tmp/asdf-nodejs-postinstall.lock.2ac8ccc4"
+declare LAST_CALLED_FILE="/tmp/asdf-nodejs-postinstall.last_called.dd05fa02"
```
These paths should be dynamic. We should probably use `mktemp` for these files.

MrQubo replied:
We need one common lock file and one common last_called file. We can check if `/tmp/asdf-nodejs-postinstall.last_called.*` exists and use the existing one if it does, generating it otherwise.
We would have to run it in the foreground thus making it a little slower.
Actually, we can even safely drop the `.2ac8ccc4` and `.dd05fa02` suffixes, it's just me being a little paranoic.

Stratus3D replied:
Yes you can get rid of those hashes if there is only going to be at most one file at a time.

Stratus3D commented:
Was this code copied from somewhere? Seems like most of it is pretty generic, except for that one line containing the reshim call. If it was copied from somewhere please cite the source so we can verify the license is compatible with ours.

Overall I think this is great. I'm curious if this could cause problems in shell scripts though. Suppose a shell script installs something and then immediately tries to use it. `asdf reshim` may have not yet finished. What does everyone else think of these changes?

MrQubo commented:
Yes, I'm aware that there will be some delay between installation and reshim. But I only thought about using in it shell prompt. I think it would cause a lot of problems with scripts like your example so don't merge it right now.

And I thought of a possibly better solution. Why not just call reshim in `asdf/shims/npm` if the second command line argument is one of `install`, `uninstall`, `update`, or the alias of one of those. It would sometimes call reshim when it's not needed, but it would always do when it's needed.
Actually, currently, `uninstall` does not work and does not cause the reshim, it'll fix this too.

The version in the first commit is based on the idea from [this gist](https://gist.github.com/niieani/29a054eb29d5306b32ecdfa4678cbb39). But I've rewritten and simplified the code in the second commit and it uses a little different idea now.

mohsenkhanpour commented:
@MrQubo @Stratus3D  If it fixes the `asdf reshim` problem on `install` and `uninstall` and it boosts the general installation speed it is worth giving it a try. It may close several issues.
How stable is this PR? I am willing to test it, if feedback is needed.

MrQubo commented:
Most asdf packages e.g. haskell, ocaml, python, elixir do not call reshim and mentions that reshim has to be called separately after installing new scripts/packages/binaries.
@Stratus3D  How about we do it similarly?
